src/tracing.c

```diff
@@ -0,0 +1,204 @@
+static voidfuncs controllers[64];
+static int       tracking_active;
+
+void InstallOpWrapper(voidfunc activate, voidfunc deactivate)
+{
+    int pos = 0;
+    while (pos < 64 && controllers[pos].activate != 0) {
+        pos++;
+    }
+    assert(pos < 64);
+    voidfuncs val = { activate, deactivate };
+    controllers[pos] = val;
+}
+
+static Obj FuncTraceInternalMethods(Obj self)
+{
+    if (tracking_active) {
+        return Fail;
+    }
+    int pos = 0;
+    while (pos < 64 && controllers[pos].activate != 0) {
+        controllers[pos].activate();
+        pos++;
+    }
+    tracking_active = 1;
+    RecordedStats = NEW_PREC(0);
+    return True;
+}
+
+static Obj FuncUntraceInternalMethods(Obj self)
+{
+    if (!tracking_active) {
+        return Fail;
+    }
+    int pos = 0;
+    while (pos < 64 && controllers[pos].deactivate != 0) {
+        controllers[pos].deactivate();
+        pos++;
+    }
+    tracking_active = 0;
+    return 0;
+}
+static Obj FuncGET_TRACED_INTERNAL_METHODS_COUNTS(Obj self)
+{
+    return RecordedStats;
+}
+
+static Obj FuncClearTraceInternalMethodsCounts(Obj self)
+{
+    RecordedStats = NEW_PREC(0);
+    return 0;
+}
+
+
+/****************************************************************************
+**
+*F * * * * * * * * * * * * * initialize module * * * * * * * * * * * * * * *
+*/
+
+/****************************************************************************
+**
+*V  GVarFuncs . . . . . . . . . . . . . . . . . . list of functions to export
+*/
+static StructGVarFunc GVarFuncs[] = {
+
+    GVAR_FUNC_0ARGS(TraceInternalMethods),
+    GVAR_FUNC_0ARGS(UntraceInternalMethods),
+    GVAR_FUNC_0ARGS(GET_TRACED_INTERNAL_METHODS_COUNTS),
+    GVAR_FUNC_0ARGS(ClearTraceInternalMethodsCounts),
+    { 0, 0, 0, 0, 0 }
+};
+
+
+/****************************************************************************
+**
+*F  InitLibrary( <module> ) . . . . . . .  initialise library data structures
+*/
+static Int InitLibrary(StructInitInfo * module)
+{
+    /* init filters and functions                                          */
+    InitGVarFuncsFromTable(GVarFuncs);
+
+    RecordedStats = NEW_PREC(0);
+    /* return success                                                      */
+    return 0;
+}
+
+/****************************************************************************
+**
+*F  InitKernel( <module> )  . . . . . . . . initialise kernel data structures
+*/
+static Int InitKernel(StructInitInfo * module)
+{
+    InitHdlrFuncsFromTable(GVarFuncs);
+    InitGlobalBag(&RecordedStats, "src/tracing.c:RecordedStats");
+    return 0;
+}
+
+
+/****************************************************************************
+**
+*F  InitInfoStats() . . . . . . . . . . . . . . . . . table of init functions
+*/
+static StructInitInfo module = {
+    // init struct using C99 designated initializers; for a full list of
+    // fields, please refer to the definition of StructInitInfo
+    .type = MODULE_BUILTIN,
+    .name = "stats",
```

Comment:
wrong! now we have two internal modules with identical name (I am actually surprised this didn't trigger an error; we ought to add a check for that; in the meantime, it can potentially break savestates

Reply:
Actually, it turns out this it's 3! (stats.c, profile.c and tracing.c all use .name = "stats")...